File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void reverse(int nums[], int start, int end) {
        while(end > start) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            
            start++;
            end--;
        }
    }

    public void rotate(int[] nums, int k) {

        k = k % nums.length;

        reverse(nums, 0, nums.length-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, nums.length-1);

        // int cy = 0;

        // for(int i = 0; i < nums.length; i++) {

        //     int prev = nums[cy];
        //     int curr = 0;
        //     int j = (cy + k) % nums.length;
        //     while(j != cy) {
        //         curr = nums[j];
        //         nums[j] = prev;
        //         prev = curr;

        //         j = (j + k) % nums.length;
        //         ++i;
        //     }

        //     nums[j] = prev;
        //     ++cy;
        // }

    }
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def reverse_list(self, nums: List[int], start: int, end : int) -> None :
        
        while start <= end :
            nums[start],nums[end] = nums[end],nums[start]
            end -= 1
            start += 1
        
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        k = k % len(nums)
        
        self.reverse_list(nums, 0, len(nums) - 1)
        self.reverse_list(nums, 0, k - 1)
        self.reverse_list(nums, k, len(nums) - 1)

